dot/sync/worker_pool.go

```diff
@@ -0,0 +1,288 @@
+	}
+
+	// check if the punishment is not valid
+	if peerSync.status == punished && peerSync.punishmentTime.Before(time.Now()) {
+		peerSync.status = available
+		s.workers[who] = peerSync
+	}
+}
+
+// submitBoundedRequest given a request the worker pool will driven it
+// to the given peer.ID, used for tip sync when we receive a block announce
+// from a peer and we want to use the exact same peer to request blocks
+func (s *syncWorkerPool) submitBoundedRequest(request *network.BlockRequestMessage,
+	who peer.ID, resultCh chan<- *syncTaskResult) {
+	s.taskQueue <- &syncTask{
+		boundTo:  &who,
+		request:  request,
+		resultCh: resultCh,
+	}
+}
+
+// submitRequest given a request the worker pool will get the very first available worker
+// to perform the request, the response will be dispatch in the resultCh
+func (s *syncWorkerPool) submitRequest(request *network.BlockRequestMessage, resultCh chan<- *syncTaskResult) {
+	s.taskQueue <- &syncTask{
+		request:  request,
+		resultCh: resultCh,
+	}
+}
+
+// submitRequests takes an set of requests and will submit to the pool through submitRequest
+// the response will be dispatch in the resultCh
+func (s *syncWorkerPool) submitRequests(requests []*network.BlockRequestMessage, resultCh chan<- *syncTaskResult) {
+	for _, request := range requests {
+		s.submitRequest(request, resultCh)
+	}
+}
+
+// punishPeer given a peer.ID we check increase its times punished
+// and apply the punishment time using the base timeout of 5m, so
+// each time a peer is punished its timeout will increase by 5m
+func (s *syncWorkerPool) punishPeer(who peer.ID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	worker, has := s.workers[who]
+	if !has {
+		return
+	}
+
+	timesPunished := worker.timesPunished + 1
+	punishmentTime := time.Duration(timesPunished) * punishmentBaseTimeout
+	logger.Debugf("⏱️ punishement time for peer %s: %.2fs", who, punishmentTime.Seconds())
+
+	s.workers[who] = &peerSyncWorker{
+		status:         punished,
+		timesPunished:  timesPunished,
+		punishmentTime: time.Now().Add(punishmentTime),
+	}
+}
+
+func (s *syncWorkerPool) ignorePeerAsWorker(who peer.ID) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	delete(s.workers, who)
+	s.ignorePeers[who] = struct{}{}
+}
+
+// totalWorkers only returns available or busy workers
+func (s *syncWorkerPool) totalWorkers() (total uint) {
+	s.mtx.RLock()
+	defer s.mtx.RUnlock()
+
+	for _, worker := range s.workers {
+		if worker.status == available {
+			total += 1
+		}
+	}
+
+	return total
+}
+
+// getAvailablePeer returns the very first peer available, if there
+// is no peer avaible then the caller should wait for availablePeerCh
+func (s *syncWorkerPool) getAvailablePeer() peer.ID {
+	for peerID, peerSync := range s.workers {
+		switch peerSync.status {
+		case punished:
+			// if the punishedTime has passed then we can
+			// use it as an available peer
+			if peerSync.punishmentTime.Before(time.Now()) {
```

Comment:
should the status be updated here?

Reply:
The status will be updated by the caller, here we just return the first peer that can be used to retrieve blocks